Add openocd

MementoRC opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| Rust            | `@conda-forge/help-rust`      |
| Go              | `@conda-forge/help-go`        |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Zulip chat][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://conda-forge.zulipchat.com

All apologies in advance if your recipe PR does not receive prompt attention.
This is a high volume repository and the reviewers are volunteers. Review times
vary depending on the number of reviewers on a given language team and may be days
or weeks. We are always looking for more staged-recipe reviewers. If you are
interested in volunteering, please contact a member of @conda-forge/core.
We'd love to have the help!
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

MementoRC committed:
Add openocd

MementoRC committed:
(wip)

MementoRC committed:
Add OpenOCD

MementoRC committed:
(ref) libhidapi

MementoRC committed:
(fix) patch for external jimtcl

MementoRC committed:
(fix) --help exits non-0

MementoRC committed:
Merge branch 'main' into openocd

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/openocd/recipe.yaml```) and found it was in an excellent condition.

MementoRC committed:
(fix) need libjaylink to test build

MementoRC committed:
(fix) verbose logs. bash-up win

MementoRC committed:
(fix) nosubmodule

MementoRC committed:
(fix) unused var for osx. automake-wrapper for win

MementoRC committed:
(fix) osx: no-strict-prototypes. win: the pain

MementoRC committed:
(fix) osx: unused-but-set. win: binutils

MementoRC committed:
(fix) osx: unused-but-set. win: m2-make

MementoRC committed:
(fix) osx: unused-but-parameter. win: m2w64

MementoRC committed:
(fix) win: touch missing exec

MementoRC committed:
Merge branch 'main' into openocd

MementoRC committed:
(fix) win: remove failing test for binaries

MementoRC committed:
(fix) win: no CFLAGS? odd

MementoRC committed:
(fix) win: no pkg.m4?

MementoRC committed:
(fix) win: m2-m4?

MementoRC committed:
(fix) win: ACLOCAL_PATH?

MementoRC committed:
(fix) win: ACLOCAL_PATH?

MementoRC committed:
(fix) win: ACLOCAL_PATH?

MementoRC committed:
(fix) win: ACLOCAL_PATH?

MementoRC committed:
(fix) win: ACLOCAL_PATH?

MementoRC committed:
(fix) win: ACLOCAL_PATH?

MementoRC committed:
(fix) win: semi-column

MementoRC committed:
(fix) win: semi-column

MementoRC committed:
(fix) win: damn you to hell

MementoRC committed:
(fix) win: damn you to hell

MementoRC committed:
(fix) win: damn you to hell

MementoRC committed:
(fix) win: :?

MementoRC committed:
(fix) win: :?

MementoRC committed:
(fix) win: :?

MementoRC committed:
(fix) win: :?

MementoRC committed:
(fix) win: damn ... what a waste of time

MementoRC committed:
(fix) win

MementoRC committed:
(fix) win

MementoRC committed:
(fix) win

MementoRC committed:
(fix) win

MementoRC committed:
(fix) win

MementoRC committed:
(fix) win

MementoRC committed:
(fix) win

MementoRC committed:
(fix) win

MementoRC committed:
(fix) win

MementoRC committed:
(fix) win

MementoRC committed:
(fix) pkg-config can't find .pc?

MementoRC committed:
(fix) pkg-config can't find .pc?

MementoRC committed:
(fix) jaylink linking: man! win is a bag full of rabid diseased rats! I'm nauseated

MementoRC committed:
(fix) jaylink linking: man! win is a bag full of rabid diseased rats! I'm nauseated

MementoRC committed:
(fix) jaylink linking: man! win is a bag full of rabid diseased rats! I'm nauseated

MementoRC committed:
(fix)

MementoRC committed:
Merge branch 'main' into openocd

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix) debug libjaylink.dll

MementoRC committed:
(fix) debug libjaylink.dll

MementoRC committed:
(fix) debug libjaylink.dll

MementoRC committed:
(fix) define win export synbols

MementoRC committed:
(fix) damn, a bit odd that we need to define __dllspec :?

MementoRC committed:
(fix) damn, a bit odd that we need to define __dllspec :?

MementoRC committed:
(fix) damn, a bit odd that we need to define __dllspec :?

MementoRC committed:
(fix) damn, a bit odd that we need to define __dllspec :?

MementoRC committed:
(fix) are there really that many masochists out there to indulge that pos OS?

MementoRC committed:
(fix) ?

MementoRC committed:
(fix) ?

MementoRC committed:
(fix) ?

MementoRC committed:
(fix) ?

MementoRC committed:
(fix) ?

MementoRC committed:
(fix) ?

MementoRC committed:
Merge branch 'main' into openocd

MementoRC committed:
(fix) ?

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
(fix)

MementoRC committed:
Merge branch 'main' into openocd

MementoRC committed:
.

MementoRC committed:
.

MementoRC committed:
.

MementoRC committed:
.

MementoRC committed:
.

MementoRC committed:
.

MementoRC committed:
.

MementoRC committed:
.

MementoRC committed:
.

MementoRC committed:
.

MementoRC committed:
(ref) add build files

MementoRC committed:
(ref) add build files

MementoRC committed:
.

MementoRC committed:
.

MementoRC committed:
.

MementoRC committed:
(fix) truncated patch?

MementoRC committed:
(fix) truncated patch?

MementoRC committed:
(fix) windows test

MementoRC committed:
(fix) simplify

wolfv commented:
Wow, a lot of work! Should I take a look at Windows? It seems like a curious error (charmap can't decode _after_ the package is already built ...).

MementoRC commented:
Sure, I was thinking to do some debug, but got carried away by the renode recipe

MementoRC committed:
Merge branch 'main' into openocd

MementoRC committed:
Merge branch 'main' into openocd

MementoRC committed:
Merge branch 'main' into openocd

MementoRC committed:
(fix) m2-binutils?

MementoRC committed:
(fix) binutils?

MementoRC committed:
(fix) ok, no aclocal path update

MementoRC committed:
(fix) autoreconf?

MementoRC committed:
(fix) find instead?

MementoRC committed:
(fix) m2-pkg-config?

MementoRC committed:
(fix) fix overflow

MementoRC committed:
(fix) fix overflow

MementoRC committed:
(fix) fix overflow

MementoRC committed:
(fix) fix overflow

MementoRC committed:
(fix) unused var

MementoRC committed:
(fix) unused var

MementoRC commented:
@wolfv Any chances to get this investigated? I have not found anything so far, it seems to be when the `.conda` file is built. @xhochy the win build fails when checking the artifacts, some pointers would be appreciated

MementoRC committed:
(ref) skip win

xhochy commented:
Please post the windows error as a PR comment the next time. I do the majority of staged-recipes review on mobile and there it is a hassle to click through Azure logs of past commits.

MementoRC committed:
(dev) reinstating win for feedback

MementoRC commented:
@xhochy Find the error below:
```
--------------------------------------
win-64\openocd-0.12.0-h0faeac5_0.conda
--------------------------------------
-- Size: 2.2MB
-- SHA256: ff8627cb62e48bfeb6868f1fbb6f0a6cb3a101d70bdca882c261671f9ee10336
-- Contents:
Traceback (most recent call last):
  File "D:\Miniforge\Scripts\inspect_artifacts-script.py", line 10, in <module>
    sys.exit(main())
             ^^^^^^
  File "D:\Miniforge\Lib\site-packages\click\core.py", line 1161, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Miniforge\Lib\site-packages\click\core.py", line 1082, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "D:\Miniforge\Lib\site-packages\click\core.py", line 1443, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Miniforge\Lib\site-packages\click\core.py", line 788, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Miniforge\Lib\site-packages\conda_forge_ci_setup\inspect_artifacts.py", line 50, in main
    list_contents(artifact, verbose=True)
  File "D:\Miniforge\Lib\site-packages\conda_package_handling\api.py", line 235, in list_contents
    return format.list_contents(in_file, verbose=verbose, components=components)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Miniforge\Lib\site-packages\conda_package_handling\conda_fmt.py", line 164, in list_contents
    _list(fn, components=components, verbose=verbose)
  File "D:\Miniforge\Lib\site-packages\conda_package_handling\streaming.py", line 82, in _list
    print("".join(lines), end="")
  File "D:\Miniforge\Lib\encodings\cp1252.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_table)[0]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeEncodeError: 'charmap' codec can't encode characters in position 57393-57394: character maps to <undefined>
"inspect_artifacts needs conda-forge-ci-setup >=4.9.3"
```

MementoRC committed:
Merge branch 'main' into openocd

MementoRC committed:
(ref) Removing libjaylink as it is not a feedstock

MementoRC committed:
Merge branch 'main' into openocd

MementoRC committed:
Merge branch 'main' into openocd

MementoRC committed:
(ref) shared jimtcl?

MementoRC committed:
(ref) forgot, no shared lib

MementoRC committed:
(dbg) Try manual inspect

MementoRC committed:
(dbg) Try manual inspect

MementoRC committed:
(dbg) try inspect_artifacts

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/openocd/recipe.yaml```) and found some lint.

Here's what I've got...


For **recipes/openocd/recipe.yaml**:

 * ❌ Selectors in comment form no longer work in v1 recipes. Instead, if / then / else maps must be used. See lines [90].


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/14364895854](https://github.com/conda-forge/conda-forge-webservices/actions/runs/14364895854). Examine the logs at this URL for more detail.</sub>

MementoRC committed:
(dbg) try inspect_artifacts

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/openocd/recipe.yaml```) and found it was in an excellent condition.

MementoRC committed:
(dbg) try inspect_artifacts

MementoRC commented:
@xhochy @wolfv Is there a way to set: `set "PYTHONIOENCODING=utf-8"` when inspecting the artifacts on windows?
When I do it within the test phase it passes (it was failing without the set ...)
```
 │ (base) %SRC_DIR%>openocd --version 
 │ Open On-Chip Debugger 0.12.0-snapshot (2025-04-09-20:46)
 │ Licensed under GNU GPL v2
 │ For bug reports, read
 │ 	http://openocd.org/doc/doxygen/bugs.html
 │ (base) %SRC_DIR%>set "PYTHONIOENCODING=utf-8" 
 │ (base) %SRC_DIR%>inspect_artifacts --all-packages 
 │ --------------------------------------
 │ win-64\openocd-0.12.0-h0faeac5_0.conda
 │ --------------------------------------
 │ -- Size: 2.2MB
...
 │ -rw-r--r-- 0/0       7486 2025-04-09 20:41:48 share/openocd/scripts/tools/memtest.tcl 
 │ -rw-r--r-- 0/0       1424 2025-04-09 20:41:48 share/openocd/scripts/tools/test_cpu_speed.tcl 
 │
 ╰─────────────────── (took 63 seconds)
  all tests passed!
```
But then the build does not complete and fails with:
```
Inspecting artifacts
--------------------------------------
win-64\openocd-0.12.0-h0faeac5_0.conda
--------------------------------------
-- Size: 2.2MB
-- SHA256: c7a70ab1f6f53b1d95c7fb37d6c2edc00490904e366c12e381875274d0a9b4c7
-- Contents:
Traceback (most recent call last):
  File "D:\Miniforge\Scripts\inspect_artifacts-script.py", line 10, in <module>
    sys.exit(main())
             ^^^^^^
...
```

MementoRC committed:
(dbg) try update pixi.toml

MementoRC committed:
(dbg) try update build_steps.sh

MementoRC committed:
(dbg) try update build_steps.sh

MementoRC committed:
(dbg) try update run_win_build.bat

MementoRC commented:
@xhochy @wolfv WHen modifying the run_build_win.bat with the setting, the inspect_artifacts passes (but of course, the linter fails). Can you take it from here?

MementoRC committed:
(mnt) revert to clean linter

MementoRC committed:
Merge branch 'main' into openocd

MementoRC committed:
Fix include issues and update script object reference

Fix include directives and update script object handling in openocd.

MementoRC committed:
Fix include statements and update script object handling

MementoRC committed:
Fix include paths and script object references in patch

MementoRC committed:
Add jimtcl include fix patch

MementoRC committed:
Patch openocd source code in build script

Add a sed command to patch openocd source code.

MementoRC committed:
Add sed and m2-sed to OpenOCD recipe dependencies

MementoRC committed:
Fix include and script object reference in jimtcl patch

Fix include statement and update script object reference in jim_get_nvp function.

MementoRC committed:
Fix calloc usage in ulink.c for correct allocation

MementoRC committed:
Fix calloc usage in ulink and ambiqmicro source files

MementoRC committed:
Refactor sed command for calloc in build.sh

MementoRC committed:
Add support for msp432 in build script

MementoRC committed:
Add max32xxx.c to sed command in build script

MementoRC committed:
(fix) C calloc

MementoRC committed:
(dbg) win dll fail

MementoRC committed:
(dbg) win dll fail

MementoRC committed:
Merge branch 'main' into openocd

MementoRC committed:
(dbg) win dll fail

MementoRC committed:
(dbg) win dll fail

MementoRC committed:
(dbg) win dll fail

MementoRC committed:
(dbg) win dll fail

MementoRC committed:
(fix) remove russian oddities

MementoRC commented:
@xhochy @wolfv Finally revisited and fixed the 'russian' non-ascii characters
